connector/src/main/java/org/geysermc/connector/network/translators/java/JavaPluginMessageTranslator.java

```diff
@@ -28,53 +28,83 @@
 import com.github.steveice10.mc.protocol.packet.ingame.server.ServerPluginMessagePacket;
 import com.google.common.base.Charsets;
+import com.nukkitx.protocol.bedrock.packet.EmotePacket;
 import org.geysermc.connector.common.AuthType;
 import org.geysermc.connector.network.session.GeyserSession;
 import org.geysermc.connector.network.translators.PacketTranslator;
 import org.geysermc.connector.network.translators.Translator;
+import org.geysermc.connector.utils.StringByteUtil;
 import org.geysermc.cumulus.Form;
 import org.geysermc.cumulus.Forms;
 import org.geysermc.cumulus.util.FormType;
 
+import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 
 @Translator(packet = ServerPluginMessagePacket.class)
 public class JavaPluginMessageTranslator extends PacketTranslator<ServerPluginMessagePacket> {
     @Override
     public void translate(GeyserSession session, ServerPluginMessagePacket packet) {
-        // The only plugin messages it has to listen for are Floodgate plugin messages
-        if (session.getRemoteAuthType() != AuthType.FLOODGATE) {
-            return;
+        // Handle plugin channels
+        switch (packet.getChannel()) {
+            case "minecraft:register":
+                session.registerDownstreamPluginChannels(StringByteUtil.bytesToStrings(packet.getData()));
+                break;
+            case "minecraft:unregister":
+                session.unregisterDownstreamPluginChannels(StringByteUtil.bytesToStrings(packet.getData()));
+                break;
+            case "floodgate:form":
+                if (session.getRemoteAuthType() == AuthType.FLOODGATE) {
+                    handleFloodgateMessage(session, packet);
+                }
+                break;
+            case "geyser:emote":
+                handleEmote(session, packet);
+                break;
+            default:
+                //TODO feature: Here we should have a callback for extensions to handle their own plugin messages
         }
+    }
 
-        String channel = packet.getChannel();
+    private void handleEmote(GeyserSession session, ServerPluginMessagePacket packet) {
+        EmotePacket emotePacket = new EmotePacket();
 
-        if (channel.equals("floodgate:form")) {
-            byte[] data = packet.getData();
+        ByteBuffer byteBuffer = ByteBuffer.wrap(packet.getData());
+        byte[] idBytes = new byte[byteBuffer.get()];
+        byteBuffer.get(idBytes);
 
-            // receive: first byte is form type, second and third are the id, remaining is the form data
-            // respond: first and second byte id, remaining is form response data
 
-            FormType type = FormType.getByOrdinal(data[0]);
-            if (type == null) {
-                throw new NullPointerException(
-                        "Got type " + data[0] + " which isn't a valid form type!");
-            }
+        emotePacket.setEmoteId(new String(idBytes, StandardCharsets.UTF_8));
```

Comment:
Looking at this do you even need to construct the ByteBuffer, can't you just do this?
```java
emotePacket.setEmoteId(new String(packet.getData(), StandardCharsets.UTF_8));
```

Reply:
Nope, we need the buffer. Its last 8 bytes are the playing entity ID